### PR TITLE
Fix a bug #3994

### PR DESCRIPTION
**Pull request by DjangoPeng**
I have fixed the bug[#3994](https://github.com/tensorflow/tensorflow/issues/3994). So, now, we can use `Variables` as gradients with the update.
Check if gradient is `tf.Variable`:
- Yes, convert it to tensor.
- No, check its type as before.

Then, remake the `grads_and_vars` after checking and converting.

### CODE REVIEW

**Comment by danmane** [comment]
@mrry Can you review this please?

File: tensorflow/python/training/optimizer.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 from tensorflow.python.ops import state_ops
 from tensorflow.python.ops import variables
 from tensorflow.python.training import slot_creator
-
+import tensorflow as tf
```
Comment (mrry) [inline tensorflow/python/training/optimizer.py]: Remove this import (as it creates a circular dependency, and the relevant method is available from `ops`).

File: tensorflow/python/training/optimizer.py [hunk 2]
```diff
@@ -282,8 +282,14 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
     # This is a default implementation of apply_gradients() that can be shared
     # by most optimizers.  It relies on the subclass implementing the following
     # methods: _create_slots(), _prepare(), _apply_dense(), and _apply_sparse().
+
+    grads_to_tensors_list = []
+    vars_list = []
     grads_and_vars = tuple(grads_and_vars)  # Make sure repeat iteration works
```
Comment (mrry) [inline tensorflow/python/training/optimizer.py]: I think this line is redundant now. (It was previously necessary in order to be able to iterate over the argument twice, when the argument was a generator expression. However, you now only iterate over it once.)

File: tensorflow/python/training/optimizer.py [hunk 3]
```diff
@@ -282,8 +282,14 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
...
     for g, v in grads_and_vars:
+      # Check if gradient is Variable. If so, convert it to tensor
+      if isinstance(g, (variables.Variable)):
+        g = tf.convert_to_tensor(g)
```
Comment (mrry) [inline tensorflow/python/training/optimizer.py]: It seems like it would be simpler (and cover more cases) to do:

```python
if g is not None:
  try:
    g = ops.convert_to_tensor_or_indexed_slices(g)
  except TypeError:
    raise TypeError(
        "Gradient must be convertible to a Tensor or IndexedSlices, or None: %s" %g)
```

File: tensorflow/python/training/optimizer.py [hunk 4]
```diff
@@ -292,6 +298,13 @@ def apply_gradients(self, grads_and_vars, global_step=None, name=None):
             "Variable must be a tf.Variable: %s" % v)
       if g is not None:
         self._assert_valid_dtypes([g, v])
+      # If all is well, append g and v to corresponding list separately 
+      grads_to_tensors_list.append(g)
+      vars_list.append(v)
```
Comment (mrry) [inline tensorflow/python/training/optimizer.py]: Instead of maintaining two lists in sync, change this to maintain a single list of tuples, called e.g. `converted_grads_and_vars` (then use `converted_grads_and_vars` in place of `grads_and_vars` in the remainder of this function).

**Comment by mrry** [comment]
Please add a test that passes a `tf.Variable` to `Optimizer.apply_gradients()` in `optimizer_test.py`.

**Comment by DjangoPeng** [comment]
@mrry  I have updated the `optimizer.py` and added the test that passes a `tf.Variable` to `Optimizer.apply_gradient()` in `optimizer_test.py`. But, I added an import of `ops`, because I have to use the `ops.control_dependencies()` to assist that convert gradient to `tf.Variable`. :(

File: tensorflow/python/training/optimizer_test.py [hunk 5]
```diff
@@ -113,6 +114,32 @@ def testNoGradients(self):
...
+        var0 = tf.Variable([1.0, 2.0], dtype=dtype)
+        var1 = tf.Variable([3.0, 4.0], dtype=dtype)
+        cost = 5 * var0 + 3 * var1
+        global_step = tf.Variable(tf.zeros([], tf.int64), name='global_step')
+        sgd_op = tf.train.GradientDescentOptimizer(3.0)
+        grads_and_vars = sgd_op.compute_gradients(cost, [var0, var1])
+        # Convert gradients to tf.Variables
+        converted_grads = [tf.Variable(tf.zeros([2], dtype)) for i in grads_and_vars]
+        convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
+
```
Comment (DjangoPeng) [inline tensorflow/python/training/optimizer_test.py]: Convert the gradient(tensor) to `tf.Variable`

File: tensorflow/python/training/optimizer_test.py [hunk 6]
```diff
@@ -113,6 +114,32 @@ def testNoGradients(self):
...
+        sgd_op = tf.train.GradientDescentOptimizer(3.0)
+        grads_and_vars = sgd_op.compute_gradients(cost, [var0, var1])
+        # Convert gradients to tf.Variables
+        converted_grads = [tf.Variable(tf.zeros([2], dtype)) for i in grads_and_vars]
+        convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
+        
+        converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
+        with ops.control_dependencies(convert_ops):
+            opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
+
```
Comment (DjangoPeng) [inline tensorflow/python/training/optimizer_test.py]: Add the `ops.control_dependencies(convert_ops):` to ensure the `converted_grads` has been initialized.

File: tensorflow/python/training/optimizer_test.py [hunk 7]
```diff
@@ -113,6 +114,32 @@ def testNoGradients(self):
...
+        cost = 5 * var0 + 3 * var1
+        global_step = tf.Variable(tf.zeros([], tf.int64), name='global_step')
+        sgd_op = tf.train.GradientDescentOptimizer(3.0)
+        grads_and_vars = sgd_op.compute_gradients(cost, [var0, var1])
+        # Convert gradients to tf.Variables
+        converted_grads = [tf.Variable(tf.zeros([2], dtype)) for i in grads_and_vars]
+        convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
+        
+        converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
+        with ops.control_dependencies(convert_ops):
```
Comment (mrry) [inline tensorflow/python/training/optimizer_test.py]: `with tf.control_dependencies(convert_ops):` should work here.

Alternatively, you could run the `convert_ops` and `opt_op` in separate steps, which might make the test clearer.

File: tensorflow/python/training/optimizer_test.py [hunk 8]
```diff
@@ -17,6 +17,7 @@
 from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
+from tensorflow.python.framework import ops
```
Comment (mrry) [inline tensorflow/python/training/optimizer_test.py]: This import shouldn't be necessary.

File: tensorflow/python/training/optimizer_test.py (hunk 7 above)
Comment (DjangoPeng) [inline tensorflow/python/training/optimizer_test.py]: Sure thing, I agree to that the separate run is better than the `control_dependencies`. I'll update it!

File: tensorflow/python/training/optimizer_test.py [hunk 9]
```diff
@@ -128,10 +127,11 @@ def testGradientsAsVariables(self):
         convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
         
         converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
-        with ops.control_dependencies(convert_ops):
-            opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
+        opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
```
Comment (DjangoPeng) [inline tensorflow/python/training/optimizer_test.py]: Remove the `control_dependencies`

File: tensorflow/python/training/optimizer_test.py [hunk 10]
```diff
@@ -128,10 +127,11 @@ def testGradientsAsVariables(self):
...
 
         tf.initialize_all_variables().run()
+        # Run convert_ops to achieve the gradietns converting
+        sess.run(convert_ops)
         # Fetch params to validate initial values
```
Comment (DjangoPeng) [inline tensorflow/python/training/optimizer_test.py]: separate run~

File: tensorflow/python/training/optimizer_test.py [hunk 11]
```diff
@@ -128,10 +127,11 @@ def testGradientsAsVariables(self):
         convert_ops = [tf.assign(converted_grads[i], gv[0]) for i,gv in enumerate(grads_and_vars)]
         
         converted_grads_and_vars = list(zip(converted_grads, [var0, var1]))
-        with ops.control_dependencies(convert_ops):
-            opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
+        opt_op = sgd_op.apply_gradients(converted_grads_and_vars, global_step)
 
         tf.initialize_all_variables().run()
+        # Run convert_ops to achieve the gradietns converting
+        sess.run(convert_ops)
```
Comment (DjangoPeng) [inline tensorflow/python/training/optimizer_test.py]: Separately run `convert_ops` and `opt_op`!~

**Comment by DjangoPeng** [comment]
@mrry I have removed the unnecessary import and `control_dependencies`

**Comment by mrry** [comment]
Great, thanks! This LGTM after the tests pass.

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by DjangoPeng** [comment]
@mrry Sounds good~BTW, what is LGTM :)

**Comment by mrry** [comment]
It means "Looks Good To Me." And now the tests have passed, so we can merge it, thanks!
